sw/device/silicon_creator/lib/otbn_util.c

```diff
@@ -24,10 +24,25 @@ void otbn_init(otbn_t *ctx) {
 
 otbn_error_t otbn_data_ptr_to_dmem_addr(const otbn_t *ctx, otbn_ptr_t ptr,
                                         uint32_t *dmem_addr_otbn) {
-  if (ptr < ctx->app.dmem_start || ptr > ctx->app.dmem_end) {
+  uintptr_t ptr_addr = (uintptr_t)ptr;
+  uintptr_t app_dmem_data_start_addr = (uintptr_t)ctx->app.dmem_data_start;
+  uintptr_t app_dmem_data_end_addr = (uintptr_t)ctx->app.dmem_data_end;
+  uintptr_t app_dmem_bss_start_addr = (uintptr_t)ctx->app.dmem_bss_start;
+  uintptr_t app_dmem_bss_end_addr = (uintptr_t)ctx->app.dmem_bss_end;
+  uintptr_t app_dmem_bss_offset = (uintptr_t)ctx->app.dmem_bss_offset;
+
+  if (app_dmem_data_start_addr <= ptr_addr &&
+      ptr_addr < app_dmem_data_end_addr) {
+    // Pointer is in the `data` section, which is at the start of DMEM
+    *dmem_addr_otbn = ptr_addr - app_dmem_data_start_addr;
+  } else if (app_dmem_bss_start_addr <= ptr_addr &&
+             ptr_addr < app_dmem_bss_end_addr) {
+    // Pointer is in the `bss` section, which is after the data section in DMEM
+    *dmem_addr_otbn = ptr_addr - app_dmem_bss_start_addr + app_dmem_bss_offset;
+  } else {
+    // Pointer is not in a valid DMEM region
```

Comment:
Not sure if I'm missing something but dmem_bss comes immediately after dmem_data, right? Can we use this to unify the checks, i.e. `app_dmem_data_start_addr <= ptr < app_dmem_bss_end_addr`, and simplify the pointer computation, i.e. `*dmem_addr_otbn = ptr_addr - app_dmem_data_start_addr`? Then we wouldn't need `dmem_bss_offset` here.

Reply:
Hah! I just wrote the same thing on the runtime code :-D

Reply:
The bss section comes directly after the data section in DMEM address space, but not in Ibex address space (because they end up in different memory sections as defined by the mask ROM linker script, i.e. the `.data.` and `.bss` section there, and there are other files' data and bss portions in between them).

For example, if the OTBN program has a 4-byte buffer called `foo` that's stored in `data` and a 512-byte buffer called `bar` that's stored in `bss`, we have something like the following in Ibex address space:
```
0x100 .data
.... <other programs' data sections> ...
0x200 _otbn_run_rsa_verify_3072_rr_modexp_dmem_data_start (app_dmem_data_start_addr)
0x200 _otbn_run_rsa_verify_3072_rr_modexp_foo
0x204 _otbn_run_rsa_verify_3072_rr_modexp_dmem_data_end (app_dmem_data_end_addr)
.... <other programs' data sections> ...
0x400 .bss
.... <other programs' bss sections> ...
0x500 _otbn_run_rsa_verify_3072_rr_modexp_dmem_bss_start (app_dmem_bss_start_addr)
0x500 _otbn_run_rsa_verify_3072_rr_modexp_bar
0x1012 _otbn_run_rsa_verify_3072_rr_modexp_dmem_bss_end (app_dmem_bss_end_addr)
```
In DMEM address space, we want to have something like:
```
0x00 .data
0x00 foo
0x20 .bss
0x20 baz
```
Note that the `.bss` section gets aligned to 32 bits there, even though we only have 4 bytes of `.data`. This depends on the OTBN linker script (which forces both `.data` and `.bss` to be 32-byte aligned) and _also_ on `.balign` directives in the OTBN program; if `baz` had a `.balign 64`, for instance, the whole bss section would start at 0x40 in DMEM address space. And from the context of the mask ROM linker script, it's impossible to know that from symbol addresses alone!

That's why the OTBN linker script records `dmem_bss_offset`; in order to know the offset _in DMEM address space_, we need to know how much space there is between `_dmem_data_start` and `_dmem_bss_start` when they're right next to each other (i.e. in the context of `otbn.ld`, not `mask_rom.ld`).

Reply:
Oh, now I understand. Makes sense, thanks.

Reply:
Thanks for the clarification @jadephilipoom!